### PR TITLE
weatherclient should use latlng

### PR DESCRIPTION
**Pull request by dsauerbrun**
weatherclient interface uses latlng now. stationcode, xcoord/ycoord are fetched via noaa's points endpoint.

I also moved the response structs into their own file to keep client.go a bit cleaner

### CODE REVIEW

File: pkg/http/controllers.go [hunk 1]
```diff
@@ -5,7 +5,7 @@ import (
 	"fmt"
 	"net/http"
 
-	"github.com/dsauerbrun/cragcast/pkg/weather-client"
+	client "github.com/dsauerbrun/cragcast/pkg/weather-client"
```
Comment (joshrosso) [inline pkg/http/controllers.go:8]: No need to explicitly set the identifier `client` here for this import declaration. With the import statement `"github.com/dsauerbrun/cragcast/pkg/weather-client"`, everything after the `-` will be the identifier. Meaning `client` will already be the identifier.

File: pkg/weather-client/client.go [hunk 2]
```diff
@@ -5,70 +5,15 @@ import (
...
-}
-
-type ForecastResponse struct {
-	Properties Properties `json:"properties"`
-	Type       string     `json:"type"`
-}
-
 type WeatherClient interface {
-	GetForecast(xCoordinate, yCoordinate float32) (*ForecastResponse, error)
+	GetForecast(lat float64, lng float64) (*ForecastResponse, error)
```
Comment (joshrosso) [inline pkg/weather-client/client.go]: **nit**: Common in go with 2 parameters of the same type to only declare the type once.

Meaning, this becomes:

```go
GetForecast(lat, lng float64) (*ForecastResponse, error)
```

File: pkg/weather-client/types.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
+package client
```
Comment (joshrosso) [inline pkg/weather-client/types.go:1]: **praise:** thanks for moving this into a separate file, looks great.

_feel free to self-resolve this conversation after reading_

File: pkg/weather-client/types.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	Updated           time.Time `json:"updated"`
+	ValidTimes        string    `json:"validTimes"`
+}
+
+type ForecastResponse struct {
+	Properties Properties `json:"properties"`
+	Type       string     `json:"type"`
+}
+
+type PointResponse struct {
```
Comment (joshrosso) [inline pkg/weather-client/types.go:59]: Are you generating these response types?

I recommend not embedding struct types unless there's a really solid reason. There are a few benefits to not embedding:

1. Reusability of "sub-structs". 
    * For example, if `Geometry` ends up getting used in more than 1 response, it doesn't need to be re-declared/maintained.
2. Docs
    * When structs aren't embedded we get to add docs (comments) to them describing what they are/for/etc. This enables IDEs to provide this context when folks are accessing them.

See [Unembed ForcastResponse struct](https://github.com/dsauerbrun/cragcast/pull/5/commits/632d4947032fba18611ec2f3f7215514a220b066) for en example of unembedding these.

File: pkg/weather-client/client.go [hunk 5]
```diff
@@ -83,10 +28,37 @@ func New() WeatherClient {
 	return newNoaaClient()
 }
 
-func (nc *NoaaClient) GetForecast(xCoordinate float32, yCoordinate float32) (*ForecastResponse, error) {
-	// BOU weather office(boulder), 52X, 75Y this is boulder
-	url := fmt.Sprintf("%s/gridpoints/BOU/%f,%f/forecast/hourly",
-		nc.APIEndpoint, xCoordinate, yCoordinate)
+func (nc *NoaaClient) getNoaaPointInfo(lat float64, lng float64) (xCoord float64, yCoord float64, stationCode string, err error) {
```
Comment (joshrosso) [inline pkg/weather-client/client.go]: **nit:** Put private methods/functions at the bottom of files so exported (public) ones show up first.

Comment (joshrosso) [inline pkg/weather-client/client.go]: Seems code would be cleaner if you returned `*PointResponse` here?

```go
func (nc *NoaaClient) getNoaaPointInfo(lat float64, lng float64) (*PointResponse, error) {
```

This would also simplify your error returns to `return nil, err`.

Most importantly, you're using **named result parameters** here, namely things like `xCoord float64` instead of just `float64`. This is a fairly uncommon practice. Named result parameters actually **initialize** those named variables at the start of the function/method. Meaning you have a variable `xCoord`, available right at the start. The idea is that you can now assign `xCoord = float64(pointData.Properties.GridX)` at some point. Then when the function exits or you call `return`, the value of `xCoord` is returned as the first return parameter. So concretely, this would change to:

```go
xCoord = float64(pointData.Properties.GridX)
yCoord = float64(pointData.Properties.GridY)
stationCode = pointData.Properties.Cwa
err = nil

// more code if there were any

return
```

In short, generally this isn't something you'll do and/or you'll know when you need it. If for any reason you want to help the client (caller) identify what the return parameters represent its best to:

* a consider if you should be returning a more clear structured type (which is my suggestion with `*PointResponse`.

  -- or --

* use documentation (comments) to describe the return parameters.

**Review by joshrosso — CHANGES_REQUESTED** [review]
@dsauerbrun want to make sure https://github.com/dsauerbrun/cragcast/issues/3 is on your radar. It could save you from some of the boilerplate work around types.

Requested changes throughout, looks like its headed in a good direction!

File: pkg/weather-client/types.go (hunk 4 above)
Comment (dsauerbrun) [reply pkg/weather-client/types.go:59]: yeah, I've been using gojsonstruct as a stopgap until I work on #3 

This is pure laziness on my part since I wanted to work on other stuff that was less boring.

File: pkg/http/controllers.go (hunk 1 above)
Comment (dsauerbrun) [reply pkg/http/controllers.go:8]: ah, I think this was automatically done by OD's linting config that I have installed on vs code.

File: pkg/weather-client/client.go (hunk 5 above)
Comment (dsauerbrun) [reply pkg/weather-client/client.go]: Gotcha, my intention was to make the function return a simple version of what we need as opposed to a vomit of information from the points endpoint.

I can just adjust the type so it doesn't unmarshall a bunch of stuff when we convert the json.

File: pkg/http/controllers.go (hunk 1 above)
Comment (dsauerbrun) [reply pkg/http/controllers.go:8]: ignoring this as it is going to be resolved in a future PR that i don't want a conflict on

File: pkg/weather-client/client.go (hunk 5 above)
Comment (dsauerbrun) [reply pkg/weather-client/client.go]: quick question, 

you're suggesting `func (nc *NoaaClient) getNoaaPointInfo(lat float64, lng float64) (*PointResponse, error) {`

why not make it return a non-pointer PointResponse? We don't need to actually pass the original right? Is it typically best practice to default to passing a pointer and only pass the copied object under specific circumstances?

EDIT: nvm, answered my own question, you can't return nil if it's not a pointer.

File: pkg/weather-client/types.go (hunk 4 above)
Comment (joshrosso) [reply pkg/weather-client/types.go:59]: Figured as much. ChatGPT will do a good job unembedding these, fwiw.

File: pkg/weather-client/client.go (hunk 5 above)
Comment (joshrosso) [reply pkg/weather-client/client.go]: You got it -- your head is in the right place. However, common convention to return nil rather than an empty-initialized struct. Also, while convention, in practice this does introduce many nil access panics to the world 🙂 when callers don't properly check/handle errors.

**Review by joshrosso — APPROVED** [review]
lgtm, with the note that https://github.com/dsauerbrun/cragcast/pull/7/files#r1179129420 will be addressed in a future PR.
